Switch box-shadows to use brush masks (and add basic brush prim).

glennw opened the pull request:
This introduces a new brush primitive type. Right now, this is
extremely basic and doesn't provide much new functionality. As
more items are ported to use it, the functionality of the brush
shader type will expand (mostly the VS parts).

The major change in this patch is to remove the old box-shadow
shader and switch box shadows to be drawn via a brush mask +
gaussian blur approach.

This *is* a performance regression for small / simple box shadows
compared to previously, but this is only temporary. The goal of
this (large) patch is to switch over to a different box-shadow
technique and address correctness issues first, and then we can
apply a number of optimizations on top of that. It *is* a performance
win on very large box shadows, which was the performance worst
case for the previous box shadow shader.

The changes included:
 * Extend blur shader to support both A8 and RGBA8 blurs.
 * Basic brush shader and primitive code.
 * Brush mask shader.
 * Remove old box shadow shader.
 * Move some common clip functions to ellipse.glsl.
 * Extend cache image shader to support both A8 and RGBA8 sources.
 * Support clip masks in cache image shader.
 * Remove box shadow primitive type.
 * Add reftests for spread, offset, blur, border radius box-shadow variations.
 * Support adjusting the border radii of box shadows as defined by spec.

Next steps:
 * Extend the mask shader to support arbitrary border radii per corner.
 * Apply various optimizations to the brush and blur code.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/1841)
<!-- Reviewable:end -->

glennw commented:
r? @kvark 

There's a couple of blockers before this can be merged - but it can probably be reviewed now.

The main issue is that the clip mask shader used for fast path box shadows (with zero blur) introduces a faint line of AA in some cases where it shouldn't. I'm a little unsure if I should leave this here until @nical finishes up his work on the AA changes, which may just resolve that problem. That causes a few test failures in the Servo WPT suite. @nical Do you think your AA work is likely to fix that, and how soon do you think it will be ready?

kvark commented:
Reviewed 30 of 30 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[webrender/res/brush.glsl, line 63 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBmAiQ_OQA0JVqnvav:-KwBmAiR3MFHdzDVYgkD:bfkx8s8) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/res/brush.glsl#L63)):*
> ```GLSL
>     // will be expanded to support transform picture types (the common kind).
>     vec2 pos = pic_task.target_rect.p0 + aPosition.xy * pic_task.target_rect.size;
>     vLocalPos = aPosition.xy * pic_task.target_rect.size / uDevicePixelRatio;
> ```

I'm concerned that this makes `vLocalPos` to live in a separate space from `vLocalRect`, which is confusing

---

*[webrender/res/cs_text_run.glsl, line 23 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBosmj8DE4KOrTl01M:-KwBosmkzPRujdrXU-RP:b1hshd7) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/res/cs_text_run.glsl#L23)):*
> ```GLSL
>     int picture_address = prim.user_data2;
> 
>     // Fetch the parent text-shadow for this primitive. This allows the code
> ```

nit: comment still implies the shadow

---

*[webrender/res/ellipse.glsl, line 70 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBoy_1tBOkwX--aZC4:-KwBoy_1tBOkwX--aZC5:b-wncfc4) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/res/ellipse.glsl#L70)):*
> ```GLSL
> 
> float clip_against_ellipse_if_needed(vec2 pos,
>                                      float current_distance,
> ```

nit: would be nicer with the new formatting style

---

*[webrender/res/ps_cache_image.glsl, line 68 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBuCu8ksvnPKHt7Hq-:-KwBuCu8ksvnPKHt7Hq0:b-vdbf8n) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/res/ps_cache_image.glsl#L68)):*
> ```GLSL
> 
>         // Apply the clip mask
>         color.a = min(color.a, do_clip());
> ```

I guess this will now become easier as @mstange convinced us to use multiplication ;)

---

*[webrender/res/ps_line.glsl, line 102 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBuTjCL1-CJzlxf1Ts:-KwBuTjCL1-CJzlxf1Tt:b5a0kwb) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/res/ps_line.glsl#L102)):*
> ```GLSL
> #ifdef WR_FEATURE_CACHE
>     int picture_address = prim.user_data0;
>     PrimitiveGeometry shadow_geom = fetch_primitive_geometry(picture_address);
> ```

should `shadow_geom` be renamed to `picture_geom`?

---

*[webrender/src/frame_builder.rs, line 1330 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBvmpt2_CAzx_pGcBb:-KwBvmpt2_CAzx_pGcBc:bpqdnr0) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/src/frame_builder.rs#L1330)):*
> ```Rust
>                         shadow_rect,
>                         BorderRadius::uniform(shadow_radius),
>                         ClipMode::ClipOut
> ```

Shouldn't this be `ClipIn`?

---

*[webrender/src/frame_builder.rs, line 1341 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBvpLyg8FoH5eJF0uW:-KwBvpLyg8FoH5eJF0uX:b-lsrwt5) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/src/frame_builder.rs#L1341)):*
> ```Rust
>                     );
> 
>                     self.add_primitive(
> ```

`add_primitive` could possibly go outside the `match`

---

*[webrender/src/frame_builder.rs, line 1431 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBwAJTkxymFktgkoE3:-KwBwAJTkxymFktgkoE4:b-nuq2iq) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/src/frame_builder.rs#L1431)):*
> ```Rust
>                     ));
> 
>                     self.add_primitive(
> ```

similarly, this could possibly be moved out of the `match`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/1841)*
<!-- Sent from Reviewable.io -->

glennw commented:
Review status: 26 of 30 files reviewed at latest revision, 8 unresolved discussions.

---

*[webrender/res/brush.glsl, line 63 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBmAiQ_OQA0JVqnvav:-KwEG2VSGH7GHen-bPsZ:beo30xj) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/res/brush.glsl#L63)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

I'm concerned that this makes `vLocalPos` to live in a separate space from `vLocalRect`, which is confusing
</blockquote></details>

Now that I look at this again, I'm not sure it correctly accounts for multiple text shadows with different offsets. I'll take a detailed look in the morning and either explain further with more comments, or fix if incorrect.

---

*[webrender/res/cs_text_run.glsl, line 23 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBosmj8DE4KOrTl01M:-KwEGEVxmbCPkFtbyw7W:b-896fix) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/res/cs_text_run.glsl#L23)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: comment still implies the shadow
</blockquote></details>

Done.

---

*[webrender/res/ellipse.glsl, line 70 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBoy_1tBOkwX--aZC4:-KwEGFKlVojjYi5CBpLg:b-896fix) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/res/ellipse.glsl#L70)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: would be nicer with the new formatting style
</blockquote></details>

Done.

---

*[webrender/res/ps_line.glsl, line 102 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBuTjCL1-CJzlxf1Ts:-KwEGJUgbFRW9lkF1_QT:b-896fix) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/res/ps_line.glsl#L102)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

should `shadow_geom` be renamed to `picture_geom`?
</blockquote></details>

Done.

---

*[webrender/src/frame_builder.rs, line 1330 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBvmpt2_CAzx_pGcBb:-KwEGKa3ByrLut7_jxBc:b-16kq4z) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/src/frame_builder.rs#L1330)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

Shouldn't this be `ClipIn`?
</blockquote></details>

I think ClipOut is named somewhat confusingly - it is the opposite of Clip (that is, it allows pixels outside the specific region).

---

*[webrender/src/frame_builder.rs, line 1341 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBvpLyg8FoH5eJF0uW:-KwEGSY_Jg6DAgf5ea1w:b-896fix) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/src/frame_builder.rs#L1341)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

`add_primitive` could possibly go outside the `match`
</blockquote></details>

Done.

---

*[webrender/src/frame_builder.rs, line 1431 at r1](https://reviewable.io:443/reviews/servo/webrender/1841#-KwBwAJTkxymFktgkoE3:-KwEGTOLSaiPhtSnI8my:b-896fix) ([raw file](https://github.com/servo/webrender/blob/718901a79af180dfc6cde97cf240dc2ce3b53be0/webrender/src/frame_builder.rs#L1431)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

similarly, this could possibly be moved out of the `match`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/1841)*
<!-- Sent from Reviewable.io -->

glennw commented:
@kvark Thanks for the review! I've addressed most of those comments. The remaining issues blocking this are:

* I need to revisit the shadow offset comment above.
* I have a local fix for the AA edge + reftest issue I mentioned above. I think I'll wait until the clip chain stuff from @mrobinson lands first though, to make sure I don't break anything there.
* I still need to do a gecko try run once I get all the Servo tests passing with the AA fix mentioned above.

nical commented on webrender/res/ellipse.glsl:
```diff
@@ -66,4 +66,56 @@ float distance_to_ellipse(vec2 p, vec2 radii) {
+                                                      vec2(1.0, -1.0),
+                                                      afwidth);
+
+    return smoothstep(0.0, afwidth, 1.0 - current_distance);
```
This implementation has a few bugs (see my ramblings in #1882), you should replace it by the corrected rounded_rect implementation from #1822.
I don't know if that's related to the aa glitch you referred to, though.

glennw replied:
Yes, the patch in #1822 fixes the test failures in this patch - it's due to the change in ```init_transform_fs```.

glennw replied:
If we land #1822 soon-ish, I'll rebase this patch on top of that.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #1855) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
Fixes #130 
Fixes #1831

Makes it trivial to implement #1430 once this lands.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #1822) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
I've rebased this over the AA change from @nical.

Gecko try run is here:
https://hg.mozilla.org/try/pushloghtml?changeset=58922e35932a38a0e9247e6f3a50845d67b02774
https://treeherder.mozilla.org/#/jobs?repo=try&revision=58922e35932a38a0e9247e6f3a50845d67b02774

There are 6 new passes, and 6 new fails.

It's likely some of these are related to other recent changes in WR (I don't think the fuzziness has been updated for the AA patch yet).

I'll do some further investigation on those failures and see if they are fuzziness issues or bugs that need to be resolved before landing this.

glennw commented:
Looking at the unexpected passes / fails:

Probably related to @Gankro 's text-shadow change:
```background-clip-text-2.html``` PASS.
```basic-opacity.html``` PASS.

Probably related to @nical 's AA change:
```preserve3d-6a.html``` FAIL (more fuzziness needed)
```backface-visibility-2.html``` FAIL (more fuzziness needed)
```split-intersect1.html``` PASS
```bugs/289480.html#top``` PASS

New box-shadow PASS from this patch:
```boxshadow-inner-basic.html``` PASS
```boxshadow-border-radius-int.html``` PASS

New box-shadow FAIL from this patch:
```box-shadow/boxshadow-large-border-radius.html``` 
```boxshadow-color-rounding-middle.html``` 
```boxshadow-inset-neg-spread.html```
```boxshadow-inset-large-offset.html```
```box-decoration-break-with-inset-box-shadow-1.html```

None of those box-shadow fails looks drastically wrong, but some of them have more fuzziness than I would *expect* to need (although most of them have fuzziness on skia / d2d too). I'll investigate some of those now.

Gankro commented:
Can confirm those two passes are what I was targeting with my patch.

glennw commented:
New try run:

https://hg.mozilla.org/try/rev/007c2de84545170b1b25b627c8a0149d6cff5350
https://treeherder.mozilla.org/#/jobs?repo=try&revision=007c2de84545170b1b25b627c8a0149d6cff5350&selectedJob=137389129

New PASSes related to this change:
```boxshadow-inner-basic.html```
```boxshadow-border-radius-int.html```

New FAILs related to this change:
```box-decoration-break-with-inset-box-shadow-1.html``` - imperceptible differences. Already marked fuzzy - needs a bit more.
```boxshadow-large-border-radius.html``` - blur has changed slightly. Already marked fuzzy - needs a bit more.
```boxshadow-inset-neg-spread.html``` - imperceptible differences. Already marked fuzzy - needs a bit more.
```boxshadow-inset-large-offset.html``` - Looks odd - already has a huge fuzzy range on WR, D2D and Skia. I think we could mark this with extra fuzziness, or as an expected FAIL for now.
 
Test differences *unrelated* to this change:
PASS - ```background-clip-text-2.html```
PASS - ```fallback-mark-stacking-1.html```
PASS - ```basic-opacity.html```
PASS - ```289480.html#top```
PASS - ```split-intersect1.html```
PASS - ```615121-2.html```
FAIL - ```preserve3d-6a.html```
FAIL - ```backface-visibility-2.html```

@staktrace @nical @kvark Given the above, I think this is probably ready to go now. Are you happy with the fuzziness changes described above?

staktrace commented:
@glennw I agree with your reftest analysis and am fine with this landing. Thanks for doing the try push!

staktrace commented:
Also for future reference, if you look at the most recent try runs in the wr-future-update bug the links to try pushes there will show which reftest differences were caused by stuff already landed and that I'm aware of. So in this case all the unrelated reftest failures that you were seeing I already have patches to adjust the fuzziness for.

kvark commented:
@staktrace @glennw thanks for being extremely cautious here!
:shipit: 🚀 🎢 
@bors-servo r+

bors-servo commented:
:pushpin: Commit 5c3b489 has been approved by `kvark`

<!-- @bors-servo r=kvark 5c3b489ec9365272c317d6f8695342ea2584e18e -->

bors-servo commented:
:hourglass: Testing commit 5c3b489ec9365272c317d6f8695342ea2584e18e with merge 6a7b5381b21805f4e3a15cda3bbe8466bc916485...

glennw commented:
@staktrace A heads up - you need to update the bindings for this enum change - https://github.com/servo/webrender/pull/1841/files#diff-937670fa1a68b9bb7dab70622773f50d - otherwise, you get serde panics when sending a display list with box shadows.

@jrmuizel Also a heads up - this does add a perf regression for box shadows. If you see large pink lines in the GPU profiler, don't worry, it's only temporary. I'll be working on the optimizations on top of this PR this week.

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/webrender/build/1.0.5868), [status-travis](https://travis-ci.org/servo/webrender/builds/289117567?utm_source=github_status&utm_medium=notification)
Approved by: kvark
Pushing 6a7b5381b21805f4e3a15cda3bbe8466bc916485 to master...

staktrace commented:
So in the [try push](https://treeherder.mozilla.org/#/jobs?repo=try&revision=2e33837af5fcbafe277ac73b35973dae3dfb987a) that included this update the reftest results are slightly different. There was no change to `boxshadow-inner-basic.html` which was supposed to be a PASS. And looking at the analyzer results I feel like `boxshadow-inset-neg-spread.html` and `boxshadow-inset-large-offset.html` should both be marked as failing as the difference between the test and reference cases is quite visible. For the former there's a light grey rect that's visible once you see the diff and know what to look for, and for the latter even though there's already a large fuzz applied, the files look visibly the same when I load them in non-WR firefox.

`boxshadow-large-border-radius.html` already had a visible difference and it's made worse but I'll fuzz it, and same with `box-decoration-with-inset-box-shadow-1.html`. In terms of gecko reftests it seems like this change overall made things worse and only fixed one reftest.
